pt/lessons/libraries/guardian.md

```diff
@@ -0,0 +1,256 @@
+
+Para atender as necessidades de todos os desenvolvedores, o Guardian implementa essa fase separadamente. Para encontrar o token os `Verify*` plugs
+
+Vamos criar alguns pipelines.
+
+```elixir
+pipeline :maybe_browser_auth do
+  plug(Guardian.Plug.VerifySession)
+  plug(Guardian.Plug.VerifyHeader, realm: "Bearer")
+  plug(Guardian.Plug.LoadResource)
+end
+
+pipeline :ensure_authed_access do
+  plug(Guardian.Plug.EnsureAuthenticated, %{"typ" => "access", handler: MyApp.HttpErrorHandler})
+end
+```
+
+Esse pipelines podem ser usados para compor diferentes requisitos de autenticação. O primeiro pipeline tenta encontrar um token primeiro na sessão e então tenta em um header. Se um é encontrado, ele vai carregar o recurso para você.
+
+O segundo pipeline exige que um token válido e verificado esteja presente e que seja do tipo "access". Para usar esses pipelones, adicione eles ao seu scope.
+
+```elixir
+scope "/", MyApp do
+  pipe_through([:browser, :maybe_browser_auth])
+
+  get("/login", LoginController, :new)
+  post("/login", LoginController, :create)
+  delete("/login", LoginController, :delete)
+end
+
+scope "/", MyApp do
+  pipe_through([:browser, :maybe_browser_auth, :ensure_authed_access])
+
+  resource("/protected/things", ProtectedController)
+end
+```
+
+As rotas de login acima vão ter o usuário autenticado se existir um. O segundo scope acima garante que um token válido é passado para todas as ações. Você não _precisa_ colocar eles em pipelines, você poderia colocá-los em seus controllers para uma customização super flexível mas nós estamos fazendo uma configuração mínima.
+
+Nós estamos esquecendo de algo até agora. O manipulador de erro adicionado no plug `EnsureAuthenticated`. Esse é um módulo muito simples que responde a
+
+* `unauthenticated/2`
+* `unauthorized/2`
+
+As duas funçoes recebem uma struct Plug.Conn e um map com paramêtros e deve lidar com os seus respectivos erros. Você pode até usar um controller do Phoenix!
+
+#### No controller
+
+Dentro do controller, há algumas opçoes de como acessar o usuário atualmente logado. Vamos começar com o mais simples.
+
+```elixir
+defmodule MyApp.MyController do
+  use MyApp.Web, :controller
+  use Guardian.Phoenix.Controller
+
+  def some_action(conn, params, user, claims) do
+    # do stuff
+  end
+end
+```
+
+Ao usar o módulo `Guardian.Phoenix.Controller`, as suas ações vão receber dois argumentos adicionais que você pode usar pattern matching. Lembre, se você não usar `EnsureAuthenticated` você pode tem user e claims nulos.
+
+A outro versão - a mais flexível/verbosa - está usando plug helpers.
+
+```elixir
+defmodule MyApp.MyController do
+  use MyApp.Web, :controller
+
+  def some_action(conn, params) do
+    if Guardian.Plug.authenticated?(conn) do
+      user = Guardian.Plug.current_resource(conn)
+    else
+      # No user
+    end
+  end
+end
+```
+
+#### Login/Logout
+
+Entrar e sair de uma sessão do browser é muito simpels. No controller de login:
```

Comment:
Since we are already using login/logout in English, I think we could use them here as well. Maybe: "Fazer o login e logout de uma sessão...". Ideas?

Reply:
I agree login/logout fits better.